src/unraphael/dash/pages/5_cluster.py

```diff
@@ -0,0 +1,298 @@
+import numpy as np
+import streamlit as st
+#from skimage import color
+from unraphael.dash.image_clustering import (align_images_to_mean, equalize_images,
+                                             cluster_images, compute_metrics,
+                                             clustimage_clustering, plot_clusters,
+                                             plot_dendrogram, matrix_of_similarities)
+from styling import set_custom_css
+from widgets import load_images_widget, show_heatmaps_widget
+#import matplotlib.pyplot as plt
+#import seaborn as sns
+#import io
+
+def show_images_widget(
+    images: dict[str, np.ndarray],
+    *,
+    n_cols: int = 4,
+    key: str = 'show_images',
+    message: str = 'Select image',
+) -> None | str:
+    """Widget to show images with given number of columns."""
+    col1, col2 = st.columns(2)
+    
+    # Ensure each number_input has a unique key
+    n_cols_key = f'{key}_cols'
+    n_cols = col1.number_input(
+        'Number of columns for display', value=4, min_value=1, step=1, key=n_cols_key
+    )    
+
+    cols = st.columns(n_cols)
+
+    for i, (name, im) in enumerate(images.items()):
+        if i % n_cols == 0:
+            cols = st.columns(n_cols)
+        col = cols[i % n_cols]
+
+        col.image(im, use_column_width=True, caption=name,clamp=True)    
+
+
+def equalize_images_widget(*, images: dict[str, np.ndarray]) -> dict[str, np.ndarray]:
+    """This widget helps with equalizing images in terms of brightness, contrast and 
+    sharpness. The mean brightness, contrast and sharpness of the images are computed"""
+    
+    col1, col2 = st.columns(2)
+
+    with col1:
+        col1.subheader('Equalization parameters')
+
+        brightness = st.checkbox('Equalize brightness', value=False)
+        contrast   = st.checkbox('Equalize contrast',   value=False)
+        sharpness  = st.checkbox('Equalize sharpness',  value=False)
+
+        preprocess_options = {
+            'brightness': brightness,
+            'contrast': contrast,
+            'sharpness': sharpness,        
+        }
+        
+        all_images = [img for name, img in images.items()]
+                
+        equalized_images = equalize_images(all_images, **preprocess_options)
+    
+    with col2:
+        metrics = compute_metrics(equalized_images)
+        
+        col2.subheader('Metrics after equalization')
+
+        col3, col4 = st.columns((2))
+        col3.metric('Mean Normalized Brightness', metrics['mean_normalized_brightness'])
+        col4.metric('SD Normalized Brightness', metrics['sd_normalized_brightness'])
+        
+        col3.metric('Mean Normalized Contrast', metrics['mean_normalized_contrast'])
+        col4.metric('SD Normalized Contrast', metrics['sd_normalized_contrast'])
+        
+        col3.metric('Mean Normalized Sharpness', metrics['mean_normalized_sharpness'])
+        col4.metric('SD Normalized Sharpness', metrics['sd_normalized_sharpness'])    
```

Comment:
I think the metrics should be a separate function that we call on a list or dict of images. We can reuse that code in the compare page.

Reply:
Good idea! It's done